### DISCUSSION TITLE
Search engine aliases

### DISCUSSION THREAD

**Discussion by make-github-pseudonymous-again** [tridactyl/tridactyl#3491] [Q&A] [answered]
How would one define a search engine called `node` such that typing `tabopen node require.resolve` would result in `tapopen g site:nodejs.org/api require.resolve` being executed?

[The request in this issue](https://github.com/tridactyl/tridactyl/issues/2142) is similar but I don't like binding. I already have search engines for python docs for instance
```
" python 3
set searchurls.py https://docs.python.org/3/search.html?check_keywords=yes&area=default&q=%s
```
and I wish to use something that behaves the same way for node. However, node does not seem to have a built-in search engine (?!). Therefore I try to exploit a generic search engine like duckduckgo to do the search. I could directly reference the exploited search engine as follows:
```
set searchurls.node https://html.duckduckgo.com/html?q=site%3Anodejs.org%2Fapi%20%s
```
But I would like to know if it is possible to write the searchul config independently of the search engine (by making an indirect reference to it). I already have the following defined:
```
set searchurls.g https://html.duckduckgo.com/html?q=%s
```
So it would be nice if I could reuse this definition instead of hardcoding the search engine. Another option would be to depend on the default search engine, but I think here it makes sense to explicitly reference a specific one that works well for doing specific site queries.

**Comment by bovine3dom** [comment, marked as answer]
Hi, there isn't currently anything like this in Tridactyl. I do something similar to your hard-coded suggestion in the example RC for Julia docs - https://github.com/tridactyl/tridactyl/blob/3d39ccb07af39cc42105e88712eb00fda687475a/.tridactylrc#L86

I'm a bit uneasy about adding the recursive expansion of search engine aliases that you want. I don't think many people would use it and it could complicate the code...

**Reply by make-github-pseudonymous-again** [reply]
Can a variable be used?

**Reply by bovine3dom** [reply]
I'm not sure what you mean.

If you're curious the code that deals with search engines is here: https://github.com/tridactyl/tridactyl/blob/df8a745f62b3c89b42998f496a2fd3240f42f664/src/lib/webext.ts#L250

**Reply by make-github-pseudonymous-again** [reply]
I mean a variable and an escaping function. Define `engine = https://html.duckduckgo.com/html?q=` and then do something like `set searchurls.node $engine+$urlencode('site:http://docs.julialang.org/en/v1 ')+%s`.

**Reply by bovine3dom** [reply]
Ah, ok. In theory we could but I'm not sure it is worth it - people would then have to escape real dollar signs, for example.

Do you have many other use cases for this that you can think of?

**Reply by make-github-pseudonymous-again** [reply]
Nope. Would be nice to have if a lot of such searchurls are used though.

**Reply by make-github-pseudonymous-again** [reply]
But I was wondering, isn't this variable/urlencode thing achievable through javascript?

**Reply by bovine3dom** [reply]
yeah, sorta, say we have the setting stored in a string called `engine` then

```
eval('`'+engine.replace(/`/g,'\\`')+'`')
```

would work. But that still introduces lots of surprises to users who thought they were writing a string but now it turns out that it's JS

**Reply by make-github-pseudonymous-again** [reply]
Ok. Can the whole `set` operation be explicitly wrapped by JS? So that no additional code is required.

**Reply by bovine3dom** [reply]
Yeah but it wouldn't be pretty:

```
jsb tri.config.set("searchengines","node",`${tri.config.get("searchurls","g")}site%3Anodejs.org%2Fapi%20%s`)
```

**Reply by bovine3dom** [reply]
(and that's obviously a one-time deal - the value of `g` will be baked in. Not such a big deal if you run it at startup in your RC, though)

**Reply by make-github-pseudonymous-again** [reply]
Thanks! Yeah that's not pretty but it works!

**Reply by make-github-pseudonymous-again** [reply]
> (and that's obviously a one-time deal - the value of `g` will be baked in. Not such a big deal if you run it at startup in your RC, though)

That's precisely my use case.
